src/metabase/query_processor/middleware/desugar.clj

```diff
@@ -4,21 +4,54 @@
    [metabase.legacy-mbql.predicates :as mbql.preds]
    [metabase.legacy-mbql.schema :as mbql.s]
    [metabase.legacy-mbql.util :as mbql.u]
+   [metabase.lib.metadata :as lib.metadata]
    [metabase.lib.util.match :as lib.util.match]
+   [metabase.query-processor.store :as qp.store]
    [metabase.util.malli :as mu]))
 
+(defn- add-missing-field-base-types
+  "Add base type to fields with id that are missing it. It is necessary for correct function
+   of [[metabase.legacy-mbql.util/desugar-is-empty-and-not-empty]]."
+  [query]
+  (lib.util.match/replace query
+    (field-clause :guard (fn [clause]
+                           (and (mbql.preds/Field? clause)
+                                (integer? (second clause))
+                                (not (contains? (get clause 2) :base-type)))))
+    (let [id (second field-clause)
+          metadata (lib.metadata/field (qp.store/metadata-provider) id)
+          {:keys [base-type]} metadata]
+      (update field-clause 2 assoc
+              :base-type base-type
+              ::desugar-added-base-type true))))
```

Comment:
This function ensures that fields with id and missing base type have base in field options prior to dusugaring. That is necessary for correct :is-empty and :not-empty desugaring.